clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java

```diff
@@ -1087,23 +1076,38 @@ private Record nextFetchedRecord() {
         }
 
         private List<ConsumerRecord<K, V>> fetchRecords(int maxRecords) {
+            // Error when fetching the next record before deserialization.
+            if (cachedRecordFetchException != null && lastRecord == null)
```

Comment:
Took me a little while to understand how this was handling fetch/parse errors. I wonder if it would be clearer to use a flag (e.g. `lastRecordFailedParsing`) to distinguish the latter case?